engine/access/rest/websockets/data_providers/events_provider.go

```diff
@@ -0,0 +1,185 @@
+	sub := p.createSubscription(subCtx)
+
+	p.BaseDataProviderImpl = NewBaseDataProviderImpl(
+		topic,
+		cancel,
+		send,
+		sub,
+	)
+
+	return p, nil
+}
+
+// Run starts processing the subscription for events and handles responses.
+//
+// No errors are expected during normal operations.
+func (p *EventsDataProvider) Run() error {
+	return subscription.HandleSubscription(p.subscription, p.handleResponse(p.send))
+}
+
+// createSubscription creates a new subscription using the specified input arguments.
+func (p *EventsDataProvider) createSubscription(ctx context.Context) subscription.Subscription {
+	if p.args.StartBlockID != flow.ZeroID {
+		return p.stateStreamApi.SubscribeEventsFromStartBlockID(ctx, p.args.StartBlockID, p.args.Filter)
+	}
+
+	if p.args.StartBlockHeight != request.EmptyHeight {
+		return p.stateStreamApi.SubscribeEventsFromStartHeight(ctx, p.args.StartBlockHeight, p.args.Filter)
+	}
+
+	return p.stateStreamApi.SubscribeEventsFromLatest(ctx, p.args.Filter)
+}
+
+// handleResponse processes an event and sends the formatted response.
+//
+// No errors are expected during normal operations.
+func (p *EventsDataProvider) handleResponse(send chan<- interface{}) func(*flow.Event) error {
+	messageIndex := counters.NewMonotonousCounter(0)
+
+	return func(event *flow.Event) error {
+		if ok := messageIndex.Set(messageIndex.Value() + 1); !ok {
+			return status.Errorf(codes.Internal, "message index already incremented to %d", messageIndex.Value())
+		}
+		index := messageIndex.Value()
+
+		send <- &models.EventResponse{
+			Event:        event,
+			MessageIndex: strconv.FormatUint(index, 10),
+		}
+
+		return nil
+	}
+}
+
+// ParseEventsArguments validates and initializes the events arguments.
+func ParseEventsArguments(
+	arguments map[string]string,
+	chain flow.Chain,
+	eventFilterConfig state_stream.EventFilterConfig,
+) (EventsArguments, error) {
+	var args EventsArguments
+
+	// Check for mutual exclusivity of start_block_id and start_block_height early
+	_, hasStartBlockID := arguments["start_block_id"]
+	_, hasStartBlockHeight := arguments["start_block_height"]
+
+	if hasStartBlockID && hasStartBlockHeight {
+		return args, fmt.Errorf("can only provide either 'start_block_id' or 'start_block_height'")
+	}
+
+	// Parse 'start_block_id' if provided
+	if hasStartBlockID {
+		var startBlockID parser.ID
+		err := startBlockID.Parse(arguments["start_block_id"])
+		if err != nil {
+			return args, fmt.Errorf("invalid 'start_block_id': %w", err)
+		}
+		args.StartBlockID = startBlockID.Flow()
+	} else {
+		args.StartBlockID = flow.ZeroID
+	}
+
+	// Parse 'start_block_height' if provided
+	if hasStartBlockHeight {
+		var err error
+		args.StartBlockHeight, err = util.ToUint64(arguments["start_block_height"])
+		if err != nil {
+			return args, fmt.Errorf("invalid 'start_block_height': %w", err)
+		}
+	} else {
+		args.StartBlockHeight = request.EmptyHeight
+	}
+
+	// Parse 'event_types' as []string{}
+	var eventTypes []string
```

Comment:
How about using the `EventTypes` type here to make the code more readable and understandable? Additionally, I believe we could use the `Parse` function for event types to validate the event type format.
https://github.com/onflow/flow-go/blob/8a3055c826a60f7a71c7a829d5d9e967c04940b0/engine/access/rest/http/request/event_type.go#L27